src/app/ReadHandler.cpp

```diff
@@ -0,0 +1,198 @@
+
+    err = ProcessReadRequest(std::move(aPayload));
+    SuccessOrExit(err);
+
+    MoveToState(HandlerState::Reportable);
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogFunctError(err);
+        Shutdown();
+    }
+
+    return err;
+}
+
+CHIP_ERROR ReadHandler::SendReportData(System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kNone));
+    SuccessOrExit(err);
+    MoveToState(HandlerState::ReportDataSending);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogFunctError(err);
+        Shutdown();
+    }
+    return err;
+}
+
+CHIP_ERROR ReadHandler::ProcessReadRequest(System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVReader reader;
+
+    ReadRequest::Parser readRequestParser;
+    EventPathList::Parser eventPathListParser;
+    TLV::TLVReader eventPathListReader;
+
+    reader.Init(std::move(aPayload));
+
+    err = reader.Next();
+    SuccessOrExit(err);
+
+    err = readRequestParser.Init(reader);
+    SuccessOrExit(err);
+
+    err = readRequestParser.CheckSchemaValidity();
+    SuccessOrExit(err);
+
+    err = readRequestParser.GetEventPathList(&eventPathListParser);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else
+    {
+        SuccessOrExit(err);
+        eventPathListParser.GetReader(&eventPathListReader);
+
+        while (CHIP_NO_ERROR == (err = eventPathListReader.Next()))
+        {
+            VerifyOrExit(TLV::AnonymousTag == eventPathListReader.GetTag(), err = CHIP_ERROR_INVALID_TLV_TAG);
+
+            EventPath::Parser eventPath;
+
+            err = eventPath.Init(eventPathListReader);
+            SuccessOrExit(err);
+            // TODO: Pass event path to report engine to generate report with interested events
+        }
+    }
+
+    // if we have exhausted this container
+    if (CHIP_END_OF_TLV == err)
+    {
+        err = CHIP_NO_ERROR;
+    }
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+const char * ReadHandler::GetStateStr() const
+{
+#if CHIP_DETAIL_LOGGING
+    switch (mState)
+    {
+    case HandlerState::Uninitialized:
+        return "Uninitialized";
+
+    case HandlerState::Initialized:
+        return "Initialized";
+
+    case HandlerState::Reportable:
+        return "Reportable";
+
+    case HandlerState::ReportDataSending:
+        return "ReportDataSending";
+    }
+#endif // CHIP_DETAIL_LOGGING
+    return "N/A";
+}
+
+void ReadHandler::MoveToState(const HandlerState aTargetState)
+{
+    mState = aTargetState;
+    ChipLogDetail(DataManagement, "IM RH moving to [%s]", GetStateStr());
+}
+
+void ReadHandler::OnReportProcessingComplete()
```

Comment:
This is never called.  Why is it here?

Reply:
it would be used in Eventing PR, and would remove it from this PR